packages/manager/apps/ips/src/pages/listing/listingContext.tsx

```diff
@@ -4,38 +4,100 @@ import React, {
   PropsWithChildren,
   createContext,
 } from 'react';
+import { useSearchParams } from 'react-router-dom';
 import { GetIpListParams } from '@/data/api';
 
 export type ListingContextType = {
   apiFilter: GetIpListParams;
   setApiFilter: React.Dispatch<React.SetStateAction<GetIpListParams>>;
-  ipToSearch: string;
-  setIpToSearch: React.Dispatch<React.SetStateAction<string>>;
+  hasNoApiFilter: boolean;
   expiredIps: string[];
   addExpiredIp: (expiredIp: string) => void;
 };
 
-export const ListingContext = createContext<ListingContextType | null>(null);
+export const ListingContext = createContext<ListingContextType>({
+  addExpiredIp: () => {},
+  apiFilter: {},
+  expiredIps: [],
+  hasNoApiFilter: true,
+  setApiFilter: () => {},
+});
+
+function cleanApiFilter(apiFilter: GetIpListParams) {
+  const result: Record<string, string> = {};
+  Object.entries(apiFilter).forEach(([key, value]) => {
+    if (value !== undefined) {
+      result[key] = value as string;
+    }
+  });
+  return result;
+}
+
+function parseSearchValue(value: string) {
+  switch (value) {
+    case 'true':
+      return true;
+    case 'false':
+      return false;
+    case 'null':
+      return null;
+    case 'undefined':
+      return undefined;
+    default:
+      return value;
+  }
+}
+
+function searchToApiFilter(search?: URLSearchParams): GetIpListParams {
+  const params: Record<string, string | number | boolean | null> = {};
+  search?.forEach?.((value, key) => {
+    params[key] =
+      key === 'version' ? parseInt(value, 10) : parseSearchValue(value);
+  });
```

Comment:
You can include this parse logic into the parseSearchValue function.
```suggestion
function parseSearchValue(key: string, value: string | number) {
  [
    {
      predicate : () => key === 'version' && ! isNaN(value),
      parse: () => parseInt(value, 10)
    },
    {
      predicate : () => value === 'true',
      parse: () => true
    },
    {
      predicate : () => value === 'false',
      parse: () => false
    },
    {
      predicate : () => value === 'null',
      parse: () => null
    },
    {
      predicate : () => value === 'undefined',
      parse: () => undefined
    } 
  ].find(({predicate}) => predicate())?.parse()?? value;
}

function searchToApiFilter(search?: URLSearchParams): GetIpListParams {
  const params: Record<string, string | number | boolean | null> = {};
  search?.forEach?.((value, key) => {
    params[key] = parseSearchValue(key, value);
  });
```

Reply:
As you can see your solution has more lines and is more complex (more operations executed) = harder to maintain

You have to traverse an array executing the predicate function on each element (5 executions) + execute the parse function if relevant or return the value if not (usual use case)

In the other solution it's 1 switch case comparing directly the value to return the right parsed value and 1 ternary to check the key